griptape/tools/google_drive/tool.py

```diff
@@ -0,0 +1,335 @@
+from google.auth.exceptions import MalformedError
+from googleapiclient.discovery import Resource
+from googleapiclient.http import MediaIoBaseUpload
+from googleapiclient.errors import HttpError
+from io import BytesIO
+
+
+@define
+class GoogleDriveClient(BaseGoogleClient, BaseTool):
+    LIST_FILES_SCOPES = ['https://www.googleapis.com/auth/drive.readonly']
+    UPLOAD_FILE_SCOPES = ['https://www.googleapis.com/auth/drive.file']
+    GOOGLE_EXPORT_MIME_MAPPING = {
+        "application/vnd.google-apps.document":
+            "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
+        "application/vnd.google-apps.spreadsheet":
+            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
+        "application/vnd.google-apps.presentation":
+            "application/vnd.openxmlformats-officedocument.presentationml.presentation"
+    }
+
+    owner_email: str = field(kw_only=True)
+
+    @activity(config={
+        "description": "Can be used to list files in a specific Google Drive folder or the root directory",
+        "schema": Schema({
+            Literal("folder_path",
+                    description="Path of the Google Drive folder (like 'MainFolder/Subfolder1/Subfolder2') "
+                                "from which files should be listed. Specify 'root' to list files from the "
+                                "root directory."): str,
+        })
+    })
+    def list_files(self, params: dict) -> ListArtifact | ErrorArtifact:
+        values = params["values"]
+
+        try:
+            service = self._build_client(self.LIST_FILES_SCOPES)
+
+            if values["folder_path"] == "root":
+                query = "mimeType != 'application/vnd.google-apps.folder' and 'root' in parents and trashed=false"
+            else:
+                folder_id = self._path_to_file_id(service, values["folder_path"])
+                if folder_id:
+                    query = f"'{folder_id}' in parents and trashed=false"
+                else:
+                    return ErrorArtifact(f"Could not find folder: {values['folder_path']}")
+
+            items = self._list_files(service, query)
+            return ListArtifact([TextArtifact(i) for i in items])
+
+        except Exception as e:
+            logging.error(e)
+            return ErrorArtifact(f"error retrieving files from Google Drive: {e}")
+
+    @activity(
+        config={
+            "description": "Can be used to save memory artifacts to Google Drive using folder paths",
+            "schema": Schema(
+                {
+                    "memory_name": str,
+                    "artifact_namespace": str,
+                    "file_name": str,
+                    Optional(
+                        "folder_path",
+                        description="Path of the Google Drive folder (like 'MainFolder/Subfolder1/Subfolder2') "
+                                    "where the file should be saved. Specify 'root' to save in the "
```

Comment:
You don't need to prompt the LLM to provide "root" as a value.

Reply:
Done!